osu.Game/Overlays/Settings/Sections/Maintenance/MigrationSelectScreen.cs

```diff
@@ -65,7 +66,11 @@ protected override void OnSelection(DirectoryInfo directory)
                         return;
                     }
 
-                    target = target.CreateSubdirectory("osu-lazer");
+                    //Check for a root directory
+                    if (target.Parent == null)
+                        target = Directory.CreateDirectory(Path.Combine(target.FullName, "osu-lazer"));
+                    else
+                        target = target.CreateSubdirectory("osu-lazer");
```

Comment:
I'm very confused. Did you test this? You've put this *inside* the `if` above which outlines the issue you're trying to fix, so I don't think this code would be hit as a result?

Would you not want something like this instead?

```diff
diff --git a/osu.Game/Overlays/Settings/Sections/Maintenance/MigrationSelectScreen.cs b/osu.Game/Overlays/Settings/Sections/Maintenance/MigrationSelectScreen.cs
index 57eb5ce60e..78100ba3c8 100644
--- a/osu.Game/Overlays/Settings/Sections/Maintenance/MigrationSelectScreen.cs
+++ b/osu.Game/Overlays/Settings/Sections/Maintenance/MigrationSelectScreen.cs
@@ -47,8 +47,7 @@ protected override void OnSelection(DirectoryInfo directory)
                 var directoryInfos = target.GetDirectories();
                 var fileInfos = target.GetFiles();
 
-                //With an empty disk (mb flash drive), this if could be false
-                if (directoryInfos.Length > 0 || fileInfos.Length > 0)
+                if (directoryInfos.Length > 0 || fileInfos.Length > 0 || target.Parent == null)
                 {
                     // Quick test for whether there's already an osu! install at the target path.
                     if (fileInfos.Any(f => f.Name == OsuGameBase.CLIENT_DATABASE_FILENAME))
@@ -67,10 +66,7 @@ protected override void OnSelection(DirectoryInfo directory)
                     }
 
                     //Check for a root directory
-                    if (target.Parent == null)
-                        target = Directory.CreateDirectory(Path.Combine(target.FullName, "osu-lazer"));
-                    else
-                        target = target.CreateSubdirectory("osu-lazer");
+                    target = target.CreateSubdirectory("osu-lazer");
                 }
             }
             catch (Exception e)

```

Reply:
The problem is in CreateSubdirectory method itself. It checks if the given folder is root and throw an exception (i wrote about this in the issue thread). The code above will exec CreateSubdirectory on any root folder, which is not what we need. Also pasting code from IO again:
```csharp
// We want to make sure the requested directory is actually under the subdirectory.
if (trimmedNewPath.StartsWith(trimmedCurrentPath, PathInternal.StringComparison)
    // Allow the exact same path, but prevent allowing "..\FooBar" through when the directory is "Foo"
    && ((trimmedNewPath.Length == trimmedCurrentPath.Length) || PathInternal.IsDirectorySeparator(newPath[trimmedCurrentPath.Length])))
{
    FileSystem.CreateDirectory(newPath);
    return new DirectoryInfo(newPath);
}

// We weren't nested
throw new ArgumentException(SR.Format(SR.Argument_InvalidSubPath, path, FullPath), nameof(path));
```

Reply:
Okay, so it was very unclear what you were actually trying to achieve here.

There needs to be inline comments explaining this, and probably should explain that we aren't using `CreateSubDirectory` due to the flaw (and always use the manual method instead).

I would still propose also fixing the outer if conditional as per my diff, rather than adding a comment explaining a remaining flaw.